src/web_socket.toit

```diff
@@ -0,0 +1,405 @@
+  next_fragment_ -> FragmentReader_?:
+    // Named block:
+    get_more := :
+      next := socket_.read
+      if next == null:
+        if pending_ == null: return null
+        throw "CONNECTION_CLOSED"
+      pending_ = pending_ ? pending_ + next : next
+
+    while pending_ == null or pending_.size < 2: get_more.call
+
+    masking := pending_[1] & 0x80 != 0
+    len := pending_[1] & 0x7f
+    header_size_needed := ?
+    if len == TWO_BYTE_SIZE_:
+      header_size_needed = masking ? 8 : 4
+    else if len == EIGHT_BYTE_SIZE_:
+      header_size_needed = masking ? 14 : 10
+    else:
+      header_size_needed = masking ? 6 : 2
+
+    while pending_.size < header_size_needed: get_more.call
+
+    if len == TWO_BYTE_SIZE_:
+      len = BIG_ENDIAN.uint16 pending_ 2
+    else if len == EIGHT_BYTE_SIZE_:
+      len = BIG_ENDIAN.int64 pending_ 2
+
+    masking_bytes := masking
+        ? pending_.copy (header_size_needed - 4) header_size_needed
+        : null
+    result := FragmentReader_ this len pending_[0] --masking_bytes=masking_bytes
+    if not result.is_ok_:
+      close
+      throw "PROTOCOL_ERROR"
+    pending_ = pending_[header_size_needed..]
+    return result
+
+  /**
+  Sends a ByteArray or string as a framed WebSockets message.
+  Strings are sent as text, whereas byte arrays are sent as binary.
+  */
+  send data -> none:
+    writer := start_sending --size=data.size
+    writer.write data
+    writer.close
+
+  /**
+  Starts sending a message to the peer.
+  If the size is given, the message is sent as a single fragment so that
+    the receiver knows the size.  In theory, proxies could split the fragment
+    into multiple fragments, which would ruin this.  In practice this seems
+    rare, and with TLS it is almost impossible.
+  The fragment size is at the WebSockets level and has no connection with
+    the size of IP packets or TLS buffers.
+  The message is sent as a text message if the first data written to the
+    writer is a string, otherwise as a binary message.
+  Returns a writer, which must be completed (all data sent, and closed) before
+    this method can be called again.
+  */
+  start_sending --size/int?=null -> WebSocketWriter:
+    if current_writer_: throw "PREVIOUS_WRITER_NOT_CLOSED"
+    current_writer_ = WebSocketWriter.private_ this size
+    return current_writer_
+
+  writer_close_ writer/WebSocketWriter -> none:
+    if current_writer_ != writer: throw "WRONG_WRITER_CALLED_CLOSE"
+    current_writer_ = null
+
+  write_ data from=0 to=data.size -> none:
+    written := 0
+    while from < to:
+      from += socket_.write data from to
+
+  reader_close_ -> none:
+    current_reader_ = null
+
+  /**
+  Abruptly closes the WebSocket.
+  Does not send a close packet first.
+  */
+  close:
+    socket_.close
+    current_writer_ = null
+    current_reader_ = null
+
+  static add_client_upgrade_headers_ headers/Headers -> string:
+    nonce := base64.encode (ByteArray 16: random 0x100)
```

Comment:
Add a note that the nonce isn't super important and doesn't need to by cryptographically safe.